### PR TITLE
Problem: mempool iteration is not thread safe

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #675

Solution:
- hold the lock during iteration, blocking concurrent modifications

A different approach than https://github.com/crypto-org-chain/cosmos-sdk/pull/681

---

## Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] provided a link to the relevant issue or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
* [ ] added a changelog entry to `CHANGELOG.md`
* [ ] updated the relevant documentation or specification, including comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] confirmed all CI checks have passed

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic, API design and naming, documentation is accurate, tests and test coverage

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -66,6 +66,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 ### Bug Fixes
 
 * (x/bank) [#20028](https://github.com/cosmos/cosmos-sdk/pull/20028) Align query with multi denoms for send-enabled.
+* (baseapp) [#]() Fix data race in mempool iteration.
```
Comment (yihuang) [inline CHANGELOG.md]:
```suggestion
* (baseapp) [#699](https://github.com/crypto-org-chain/cosmos-sdk/pull/699) Fix data race in mempool iteration.
```

**Comment by mmsqe** [comment]
SenderNonceMempool seems still have issue

**Comment by yihuang** [comment]
> SenderNonceMempool seems still have issue

what issue?

**Comment by mmsqe** [comment]
> > SenderNonceMempool seems still have issue
> 
> what issue?

no now, error from cb is handled

**Review by mmsqe — APPROVED** [review]

File: baseapp/abci_utils.go [hunk 2]
```diff
@@ -284,14 +284,18 @@ func (h *DefaultProposalHandler) PrepareProposalHandler() sdk.PrepareProposalHan
...
+		var (
+			resError        error
+			selectedTxsNums int
+			invalidTxs      []sdk.Tx // invalid txs to be removed after the iteration
+		)
+		h.mempool.SelectBy(ctx, req.Txs, func(memTx mempool.Tx) bool {
 			signerData, err := h.signerExtAdapter.GetSigners(memTx.Tx)
 			if err != nil {
-				return nil, err
+				// propogate the error to the caller
```
Comment (yihuang) [inline baseapp/abci_utils.go]:
```suggestion
				// propagate the error to the caller
```

**Pull request merged**